File: backend/app/schemas/highlight.py
# Pydantic schemas for highlight creation and highlight API responses.

from pydantic import BaseModel, Field
from typing import List, Optional


# Payload for creating a text highlight.
class TextHighlightCreateRequest(BaseModel):
    paper_id: int
    paragraph_id: Optional[int] = None
    scope: str
    field_name: str
    item_index: Optional[int] = None
    language: str = Field(..., pattern="^(en|zh)$")
    start_offset: int
    end_offset: int
    color: str


# Response payload for a stored text highlight.
class TextHighlightResponse(BaseModel):
    id: int
    paper_id: int
    paragraph_id: Optional[int]
    scope: str
    field_name: str
    item_index: Optional[int]
    language: str
    start_offset: int
    end_offset: int
    color: str


# Payload for creating a PDF highlight.
class PdfHighlightCreateRequest(BaseModel):
    paper_id: int
    paragraph_id: Optional[int] = None
    page_number: int
    rects: List[List[float]]
    color: str


# Response payload for a stored PDF highlight.
class PdfHighlightResponse(BaseModel):
    id: int
    paper_id: int
    paragraph_id: Optional[int]
    page_number: int
    rects: List[List[float]]
    color: str


# Combined text and PDF highlight response for one paper.
class PaperHighlightsResponse(BaseModel):
    text_highlights: List[TextHighlightResponse]
    pdf_highlights: List[PdfHighlightResponse]

File: frontend/src/types/highlight.ts
// Frontend highlight types shared across highlight UI and API clients.

// Allowed highlight colors shared by text and PDF highlights.
export type HighlightColor = "yellow" | "green" | "pink";

// Stored text highlight represented by offsets into a rendered field.
export type TextHighlight = {
  id: number;
  paper_id: number;
  paragraph_id: number | null;
  scope: "paragraph" | "overview";
  field_name: string;
  item_index: number | null;
  language: "en" | "zh";
  start_offset: number;
  end_offset: number;
  color: HighlightColor;
};

// Stored PDF highlight represented by normalized page rectangles.
export type PdfHighlight = {
  id: number;
  paper_id: number;
  paragraph_id: number | null;
  page_number: number;
  rects: number[][];
  color: HighlightColor;
};

// Combined highlight response for one paper.
export type PaperHighlightsResponse = {
  text_highlights: TextHighlight[];
  pdf_highlights: PdfHighlight[];
};

// Request payload for creating a text highlight.
export type CreateTextHighlightPayload = {
  paper_id: number;
  paragraph_id?: number | null;
  scope: "paragraph" | "overview";
  field_name: string;
  item_index?: number | null;
  language: "en" | "zh";
  start_offset: number;
  end_offset: number;
  color: HighlightColor;
};

// Request payload for creating a PDF highlight.
export type CreatePdfHighlightPayload = {
  paper_id: number;
  paragraph_id?: number | null;
  page_number: number;
  rects: number[][];
  color: HighlightColor;
};